### PR TITLE
Coral binary

### PR DESCRIPTION
**Pull request by ethanfrey**
Configure `wasmd`/`wasmcli` binaries for the coral testnet.

We use the same code, just set a few flags in the Makefile to set a unique bech32 prefix, directory, and node names.

Try `make build-coral` and try it out - eg look at keys, see config is different than wasm, look at help

Anything else I need to customize?

- Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/gaia/blob/master/CONTRIBUTING.md#pr-targeting))

- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote tests
- [ ] Updated relevant documentation (`docs/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

**Comment by orkunkl** [comment]
A suggestion: instead of merging to master, we can create a branch `testnets/coral` and apply changes there, thus provide a binary via `testnet-release`?

**Comment by ethanfrey** [comment]
`testnet-release` is gone as of #221 

I would prefer to limit branches, as they are a headache to maintain. Which version of corald corresponds to which version of wasmd and how to tag?

My attempt was to add multiple binaries without multiple `cmd` packages via the Makefile. This probably looks cleaner after #221

File: Makefile [hunk 1]
```diff
@@ -69,15 +83,23 @@ all: install lint test
...
 	go build -mod=readonly $(BUILD_FLAGS) -o build/wasmcli ./cmd/wasmcli
 endif
 
+build-coral: go.sum
+ifeq ($(OS),Windows_NT)
+    # wasmd nodes not supported on windows, maybe the cli?
+	go build -mod=readonly $(CORAL_BUILD_FLAGS) -o build/coral.exe ./cmd/wasmcli
+else
+	go build -mod=readonly $(CORAL_BUILD_FLAGS) -o build/corald ./cmd/wasmd
+	go build -mod=readonly $(CORAL_BUILD_FLAGS) -o build/coral ./cmd/wasmcli
```
Comment (alpe) [inline Makefile:100]: Note: I was expecting some `coralcli` like `gaiacli` or `wasmcli`

Comment (ethanfrey) [reply Makefile:100]: I thought that, but the naming is pretty bad. What other unix commands use that naming?

`sshd` -> `ssh`
`mongod` -> `mongo`
`psqld` -> `psql`
...

I do see `wasm` as a poor command name, so I will leave it, but for testnets `coral`/`corald` seems nicer to use.

Comment (ethanfrey) [reply Makefile:100]: I am a bit tired of following all the conventions of the sdk

Comment (alpe) [reply Makefile:100]: np. Just wanted to highlight this so that we can address this is the testnet doc.

Comment (ethanfrey) [reply Makefile:100]: Good to make this clear (also the reasoning)

Comment (ethanfrey) [reply Makefile:100]: Can I get an approval and merge? Or are there other issues?

**Comment by ethanfrey** [comment]
Cool, thank you fro the detailed testing.

I did much of that, but did not `Cosmos addr in contract decoding`. Cool to get the detailed checks.

**Comment by ethanfrey** [comment]
A bit out of context, but what do you think of this error message @alpe @webmaster128 

`wasm contract failed: Error calling the VM: Calling external function through FFI: Unknown error during FFI call: Some(\"invalid Bech32 prefix; expected coral, got cosmos\")`

It contains the original error message and all context, but is maybe a bit too verbose. There is a bit of a stutter there. I guess this would be more ideal:

`wasm contract failed: Error calling the VM: Unknown error during FFI call: \"invalid Bech32 prefix; expected coral, got cosmos\"`

We can make a (not so high priority) issue on cosmwasm if you like that or have idea for a better one

**Comment by alpe** [comment]
> A bit out of context, but what do you think of this error message @alpe @webmaster128
> 
> `wasm contract failed: Error calling the VM: Calling external function through FFI: Unknown error during FFI call: Some(\"invalid Bech32 prefix; expected coral, got cosmos\")`
> 
> It contains the original error message and all context, but is maybe a bit too verbose. There is a bit of a stutter there. I guess this would be more ideal:
> 
> `wasm contract failed: Error calling the VM: Unknown error during FFI call: \"invalid Bech32 prefix; expected coral, got cosmos\"`
> 
> We can make a (not so high priority) issue on cosmwasm if you like that or have idea for a better one

We can always aim for better UX but let's confirm first this is a real issue for users. It did not hurt :-)

**Comment by webmaster128** [comment]
This error will change completely when passed back to the contract (https://github.com/CosmWasm/go-cosmwasm/issues/124) and propagates as a StdError instead of a VmError.

**Comment by ethanfrey** [comment]
Good point @webmaster128 

Let's review these again when all last changes are made.

**Comment by gelinger777** [comment]
@ethanfrey  I am a bit confused. I have in linux corald in command line but not coral. 

And in your docs to setup enviornment here 
https://docs.cosmwasm.com/getting-started/setting-env.html

you have 

```
coral config chain-id $CHAIN_ID
coral config trust-node true

``` 

and coral not found . 


Should I replace in that doc all coral to corald?

**Comment by ethanfrey** [comment]
Yeah, please add an issue to docs

With the statgate upgrade, we merged them into one binary to track the sdk patterns. That should always be corald, wasmd, etc for both client and server now.

**Comment by ethanfrey** [comment]
Also config command no longer works in stargate (this is due to sdk changes) and those must be passed down as flags.

The docs are correct up to v0.11. @orkunkl can you split those docs into 2 sections for launchpad and stargate?

**Comment by gelinger777** [comment]
can you hint here how to   use the testchain ? 

what will be equivalent of this code in docs? 


```
coral config chain-id $CHAIN_ID
coral config trust-node true

# if connecting to local node: coral config node http://localhost:26657
coral config node $RPC
coral config output json
coral config indent true

# this is important, so the cli returns after the tx is in a block,
# and subsequent queries return the proper results
coral config broadcast-mode block

# check you can connect
coral query supply total
coral query staking validators
coral query wasm list-code

# add more wallets for testing
coral keys add fred
>
{
  "name": "fred",
  "type": "local",
  "address": "coral1avdvl5aje3zt0uay40uj6l9xtqtlqhduu84nql",
  "pubkey": "coralpub1addwnpepqvcjveqepq34x59fnmygdy58ag7zwu8gefgsprq9th38nxzptpgszc3rkve",
  "mnemonic": "hobby bunker rotate piano satoshi planet network verify else market spring toward pledge turkey tip slim word jaguar congress thumb flag project chalk inspire"
}

coral keys add bob
coral keys add thief

```

**Comment by alpe** [comment]
@gelinger777 `coralnet` was deprecated and is not running anymore AFAIK. I will add an issue cleanup the Makefile.
If you would join our `#testnet` channel on [discord](https://discord.com/invite/cPjEnPd) we can help you get up to speed with [musselnet](https://github.com/CosmWasm/testnets/tree/master/musselnet)

**Pull request merged**